pkg/providers.bzl

```diff
@@ -29,3 +28,69 @@ PackageVariablesInfo = provider(
         "values": "Dict of name/value pairs",
     },
 )
+
+# For the below, attributes always look something like this:
+#
+# ```
+#   {"unix": ["0755", "root", "root"]}
+# ```
+#
+
+# TODO: should all of these be singluar or plural?
+#
+# Seems like there's no harm in specifying a plural, even in rules: it would
+# allow some degree of simplification of rules without needing to specify a
+# custom manifest.
```

Comment:
I like this idea myself.  WDYT?

Reply:
I actually don't understand. I don't see how the plural-ness of the name impacts the rule. Or maybe I misunderstand the point entirely.

Reply:
The singular vs plural issue does indeed impact the rule interface.  If, for example, the directory provider is singular, then you need to have to rules like this:

```python
pkg_mkdir(name = 'a', dir_name = 'a')
pkg_mkdir(name = 'b', dir_name = 'b')
```

Since `PackageDirectoryInfo` can only contain information on one directory.   A rule that creates multiple directories simultaneously would have to return a `PackageFilegroupInfo` provider in this scheme.

If we have a plural `PackageDirectoriesInfo` provider, then a rule like

```python
pkg_mkdirs(name = 'dirs', dirs = ['a', 'b'])
```

would return one `PackageDirectoriesInfo` provider, which can then be included in a `pkg_filegroup` (or equivalent) rule.

I'm in favor of this because it allows for rules to be specified more cleanly.  Does this make sense?

Reply:
Thanks. I understand now. 
Do you think it save enough space in practice to justify this one being different?

The most frequent time I have wanted a mkdir is if I want to explicitly set the attributes of a folder that would otherwise be created with a default.
```
pkg_filegroup(
   name=docs,
   srcs=glob([docs/**]) + ["docs_dir"],
   strip_prefix='docs', path='/usr/share/MY_APP/docs',
   attrs = '0444',
)

pkg_mkdir(name="docs_dir", dir="/usr/share/MY_APP/docs", attrs="0555", owner='man', group='man')
```

So, I would not be creating many directories at once, but rather one per logical group of files.
Do you have any real-world use cases that I am missing?

Reply:
Rereading your words "all of these"   Did you imply for anything else beyond,  mkdir and symlinks?
FWIW, my thoughts on mkdir would also apply to symlink.  I would want a pkg_filegroup that represents installing the binary in /usr/share/MY_APP/bin  and creating a symlink from /usr/bin/MY_APP to /usr/share/MY_APP/bin. Then I can talk about the pair as a nice unit.

Reply:
First comment first:

> So, I would not be creating many directories at once, but rather one per logical group of files.
> Do you have any real-world use cases that I am missing?

The only cases would perhaps be when you have to apply custom permissions to entire custom directory structures.  Doing grouping like this would save users having to write many `pkg_mkdir` rules scattered about the BUILD file, and instead have a cluster of centralized rules that specifies custom permissions/ownership.

```
pkg_mkdirs(
    name = "core_dirs",
    dirs = [
        "/opt/MYAPP",
        "/opt/MYAPP/bin",
        "/opt/MYAPP/etc",
        "/opt/MYAPP/etc/MYAPP.d",
        ...
    ],
    attrs = {"unix": ("0755", "MYAPP", "MYAPP")},
)
```

FWIW, I do like the idea of pairing, though, but it looks like it might introduce a lot of boilerplate.  If the providers are plural, it would give flexibility in specifying targets and would not preclude the pairing pattern.

I vote for making providers plural if only to support additional flexibility for rule authors.

> Rereading your words "all of these" Did you imply for anything else beyond, mkdir and symlinks?

I was thinking it would apply to everything, now and in the future.

> FWIW, my thoughts on mkdir would also apply to symlink. I would want a pkg_filegroup that represents installing the binary in /usr/share/MY_APP/bin and creating a symlink from /usr/bin/MY_APP to /usr/share/MY_APP/bin. Then I can talk about the pair as a nice unit.

I like the sound of this too.

---

Ultimately, the question is about the providers, and I don't think we should be creating such restrictions.  We could, perhaps, create rules that only create one directory or one symlink, and have the providers only contain a single item in their corresponding list/dict.

Reply:
We want filegroup to be able to glob() files, so that has to have a provider with a list of files. We can't glob() symlinks and directories, because they do not exist in the source tree.  So I think of them differently, but your example is a good case for mkdirs.

OTOH, your last example raised a new design question.
Am I allowed to mix pkg_mkdir destinations with implicit directories from a pkg_filegroup? 
For example, I add this:
```
pkg_filegroup(
   name = "core_bin",
   srcs = [":my_bin"],
   prefix = "/opt/MYAPP/bin",
)

pkg_tar(
   name = "MYAPP.tgz",
   srcs = [
      ":core_bin",
      ":core_dirs",
   ]
)
```

Do we intend that tar should sort the destinations and do the right thing? I believe the answer is yes.
And also, that if `core_bin` also defines the output path /opt/MYAPP/bin explicitly, that is an error.

Reply:
> We want filegroup to be able to glob() files, so that has to have a provider with a list of files. We can't glob() symlinks and directories, because they do not exist in the source tree. So I think of them differently, but your example is a good case for mkdirs.

Another way to think about this would be to describe directories and symlinks can as "data", and adding a rule to adding multiple data entries at the same time would make sense.

WDYT?  Again, we can finish this up in a separate change.

> Do we intend that tar should sort the destinations and do the right thing? I believe the answer is yes.

Agreed.  This was one of my plans initially, I don't think I ever got around to it.

> And also, that if core_bin also defines the output path /opt/MYAPP/bin explicitly, that is an error.

Not sure what you mean here.  Do you mean a duplicate file/directory?  If so, I also agree.

Reply:
Yes. duplicate files.
symlinks may want to be singular so it can point to a label. as discussed off this PR.

Reply:
Done.  I don't think the symlink provider is quite right (I'll make another comment for that discussion), but in the name of having something mostly working and in the tree, this should be good to go.